### ISSUE TITLE
setup: MySQL SSL not used

### ISSUE THREAD

**Issue by jesdynf** [openemr/openemr#2670]
OpenEMR does not use an SSL connection to the database during initial setup.

Correct behavior:

If the certificate exists, it should be used.
If the certificate doesn't exist, proceed without encryption.
If the certificate is invalid or cannot be read (not absent, but any other error), configuration should fail without transmitting credentials.

(The third case relates to how the bug was discovered; ownership errors on my part made the file present but unreadable. Skipping the unreadable file and proceeding without encryption would not be the correct decision.)

**Comment by edcaracas** [comment]
Hi, can i work on this?

**Comment by bradymiller** [comment]
hi @edcaracas , definitely. it is yours :)

**Comment by edcaracas** [comment]
Do you have any idea on how to should be arrange into the form?
It could be a checkbox input and a file input.
![Screen Shot 2020-03-05 at 02 09 41](https://user-images.githubusercontent.com/6628306/75960981-b3006e80-5e86-11ea-87db-8971c6020a12.png)

**Comment by edcaracas** [comment]
By "certificate exists" you mean default certificates? ej. "/etc/ssl/certs/ca-certificates.crt"

**Comment by tywrenn** [comment]
@edcaracas Any way that makes it easier for system admins to setup OpenEMR is the best way to do this which would be having a checkbox that tells the setup to use SSL for mysql and then the setup shows file inputs to upload the SSL info to the proper location. @bradymiller I'm not sure if file inputs for SSL causes issues regarding security so correct me if i'm wrong.

**Comment by edcaracas** [comment]
@tywrenn Encrypted connection are handled by default for the client(MySQL prior to versions 5.7.3): https://dev.mysql.com/doc/refman/5.7/en/using-encrypted-connections.html#using-encrypted-connections-client-side-configuration

This resolves the first two mentions of the issue:
```
- If the certificate exists, it should be used.
- If the certificate doesn't exist, proceed without encryption.
```

For this mention:
`- If the certificate is invalid or cannot be read (not absent, but any other error), configuration should fail without transmitting credentials.`
I think it should be treated as a feature instead of a bug as this will require to update existing documentation.

**Comment by bradymiller** [comment]
hi @edcaracas and @tywrenn ,

To get an idea how this works in OpenEMR, check out the following code:
below is where it checks for existence of the sites/default/documents/certificates/mysql-ca file (if, so then it will use ssl)
https://github.com/openemr/openemr/blob/master/library/sql.inc#L33-L40

below is where it then decides whether to also support client side certificate/key if the sites/default/documents/certificates/mysql-key and sites/default/documents/certificates/mysql-cert exist
https://github.com/openemr/openemr/blob/master/library/ADODB_mysqli_mod.php#L53-L79

The problem is that this mechanism is not used during the setup.php/installation script. So if the cert or keys are bad then it will install ok, but will never work after install. So, the goal is to also support this same mechanism during setup.php/installation, thus if the cert or keys are bad then it will break while trying to install.

At this time, the user needs to place the cert/keys in the sites/default/documents/certificates/ directory, but adding a way to bring these in on setup.php gui would be a neat addition (but definitely considered optional to complete this issue.

**Comment by bradymiller** [comment]
@jesdynf 
Have a fix for this at #3785 
It with fail if the cert doesn't match; with white screen of death with following error in php error log:
```
SSL operation failed with code 1 ...
```
Not a nice fail, but a fail nonetheless :)

**Comment by stephenwaite** [comment]
take down that old issue!
![giphy](https://user-images.githubusercontent.com/5618372/88747273-70cd1780-d11c-11ea-84c8-040895a0500a.gif)
